src/Microsoft.AspNetCore.Dispatcher/DispatcherValueEndpointSelector.cs

```diff
@@ -2,54 +2,244 @@
-                var endpoint = context.Endpoints[i] as IDispatcherValueSelectableEndpoint;
-                if (!CompareRouteValues(dispatcherFeature.Values, endpoint.Values))
+                case 0:
+                    _logger.NoEndpointsMatched(dispatcherFeature.Values);
+                    break;
+
+                case 1:
+                    _logger.EndpointMatched(candidates[0].DisplayName);
+                    break;
+
+                default:
+                    var endpointNames = string.Join(
+                            Environment.NewLine,
+                            candidates.Select(a => a.DisplayName));
+                    _logger.AmbiguousEndpoints(endpointNames);
+                    break;
+            }
+
+            foreach (var candidate in candidates)
+            {
+                _context.Endpoints.Add(candidate);
+            }
+
+            return _context.InvokeNextAsync();
+        }
+
+        public IReadOnlyList<Endpoint> SelectCandidates(DispatcherValueCollection dispatcherValueCollection)
+        {
+            // The Cache works based on a string[] of the route values in a pre-calculated order. This code extracts
+            // those values in the correct order.
+            var keys = _cache.RouteKeys;
+            var values = new string[keys.Length];
+            for (var i = 0; i < keys.Length; i++)
+            {
+                dispatcherValueCollection.TryGetValue(keys[i], out object value);
+
+                if (value != null)
                 {
-                    context.Endpoints.RemoveAt(i);
+                    values[i] = value as string ?? Convert.ToString(value);
                 }
             }
-            
-            return context.InvokeNextAsync();
+
+            if (_cache.OrdinalEntries.TryGetValue(values, out var matchingRouteValues) ||
+                _cache.OrdinalIgnoreCaseEntries.TryGetValue(values, out matchingRouteValues))
+            {
+                Debug.Assert(matchingRouteValues != null);
+                return matchingRouteValues;
+            }
+
+            return EmptyEndpoints;
         }
-        
-        private bool CompareRouteValues(DispatcherValueCollection values, DispatcherValueCollection requiredValues)
+
+        private class Cache
         {
-            foreach (var kvp in requiredValues)
+            public Cache(IReadOnlyList<Endpoint> endpoints)
             {
-                if (string.IsNullOrEmpty(kvp.Value.ToString()))
+                // We need to build two maps for all of the route values.
+                OrdinalEntries = new Dictionary<string[], List<SimpleEndpoint>>(StringArrayComparer.Ordinal);
+                OrdinalIgnoreCaseEntries = new Dictionary<string[], List<SimpleEndpoint>>(StringArrayComparer.OrdinalIgnoreCase);
+
+                // We need to first identify of the keys that action selection will look at (in route data). 
+                // We want to only consider conventionally routed actions here.
```

Comment:
Some of these comments need to be cleaned up a bit - these refer to some MVC concepts that aren't part of the dispatcher